File: src/notify.ts
import { spawn } from 'node:child_process';
import process from 'node:process';

/**
 * Show an OS desktop notification.
 *
 * Platform-specific dispatch:
 *   - Linux: notify-send
 *   - macOS: osascript (display notification) or terminal-notifier
 *   - Windows: PowerShell toast notification
 *
 * Falls back to printing to stderr if no notification command is available.
 * Uses spawn (non-blocking) and catches all exceptions so it
 * never crashes the caller.
 */
export function notify(title: string, message: string): void {
  const platform = process.platform;
  try {
    if (platform === 'linux') {
      _notifyLinux(title, message);
      return;
    } else if (platform === 'darwin') {
      _notifyMacOS(title, message);
      return;
    } else if (platform === 'win32') {
      _notifyWindows(title, message);
      return;
    }
  } catch {
    // Fall through to stderr fallback
  }

  // Final fallback: print to stderr for debugging
  console.error(`[cc-pomodoro] ${title}: ${message}`);
}

function _notifyLinux(title: string, message: string): void {
  try {
    spawn('notify-send', [title, message], {
      stdio: 'ignore',
    }).unref();
  } catch {
    throw new Error('notify-send not found');
  }
}

function _notifyMacOS(title: string, message: string): void {
  const safeTitle = title.replace(/"/g, '\\"');
  const safeMessage = message.replace(/"/g, '\\"');

  try {
    spawn('terminal-notifier', ['-title', title, '-message', message], {
      stdio: 'ignore',
    }).unref();
    return;
  } catch {
    // Fall back to osascript
  }

  try {
    const script = `display notification "${safeMessage}" with title "${safeTitle}"`;
    spawn('osascript', ['-e', script], {
      stdio: 'ignore',
    }).unref();
    return;
  } catch {
    throw new Error('No notification command found on macOS');
  }
}

function _notifyWindows(title: string, message: string): void {
  const safeTitle = title.replace(/"/g, '`"').replace(/\$/g, '`$');
  const safeMessage = message.replace(/"/g, '`"').replace(/\$/g, '`$');

  // Method 1: PowerShell WinRT toast
  const winrtScript = [
    `[Windows.UI.Notifications.ToastNotificationManager,`,
    ` Windows.UI.Notifications, ContentType = WindowsRuntime] > $null;`,
    ` $template = [Windows.UI.Notifications.ToastNotificationManager]::`,
    `GetTemplateContent([Windows.UI.Notifications.ToastTemplateType]::`,
    `ToastText02);`,
    ` $textNodes = $template.GetElementsByTagName("text");`,
    ` $textNodes.Item(0).AppendChild(`,
    `$template.CreateTextNode("${safeTitle}")) > $null;`,
    ` $textNodes.Item(1).AppendChild(`,
    `$template.CreateTextNode("${safeMessage}")) > $null;`,
    ` $toast = [Windows.UI.Notifications.ToastNotification]::new($template);`,
    ` [Windows.UI.Notifications.ToastNotificationManager]::`,
    `CreateToastNotifier().Show($toast)`,
  ].join('');

  // Method 2: msg.exe
  const msgScript = `${title}: ${message}`;

  // Method 3: COM Popup via PowerShell
  const popupScript =
    `(New-Object -ComObject WScript.Shell).Popup(` +
    `"${safeMessage}", 5, "${safeTitle}", 0x40)`;

  // Try methods in order
  _tryWindowsMethods([winrtScript, msgScript, popupScript], safeTitle, safeMessage);
}

function _tryWindowsMethods(
  methods: string[],
  title: string,
  message: string,
): void {
  const [first, ...rest] = methods;

  if (!first) {
    throw new Error('All Windows notification methods failed');
  }

  try {
    if (methods.length === 3) {
      // WinRT toast
      spawn(
        'powershell',
        ['-NoProfile', '-NonInteractive', '-Command', first],
        {
          stdio: 'ignore',
          windowsHide: true,
        },
      ).unref();
    } else if (methods.length === 2) {
      // msg.exe
      spawn('msg', ['*', `${title}: ${message}`], {
        stdio: 'ignore',
        windowsHide: true,
      }).unref();
    } else {
      // COM Popup
      const popupScript =
        `(New-Object -ComObject WScript.Shell).Popup(` +
        `"${message}", 5, "${title}", 0x40)`;
      spawn(
        'powershell',
        ['-NoProfile', '-NonInteractive', '-Command', popupScript],
        {
          stdio: 'ignore',
          windowsHide: true,
        },
      ).unref();
    }
    return;
  } catch {
    if (rest.length > 0) {
      _tryWindowsMethods(rest, title, message);
    } else {
      throw new Error('All Windows notification methods failed');
    }
  }
}

File: cc_pomodoro/notify.py
from __future__ import annotations

import shutil
import subprocess
import sys


def notify(title: str, message: str) -> None:
    """Show an OS desktop notification.

    Platform-specific dispatch:
      - Linux: notify-send
      - macOS: osascript (display notification) or terminal-notifier
      - Windows: PowerShell toast notification

    Falls back to printing to stderr if no notification command is available.
    Uses subprocess.Popen (non-blocking) and catches all exceptions so it
    never crashes the caller.
    """
    platform = sys.platform
    try:
        if platform in ("linux", "linux2"):
            _notify_linux(title, message)
            return
        elif platform == "darwin":
            _notify_macos(title, message)
            return
        elif platform == "win32":
            _notify_windows(title, message)
            return
    except Exception:
        pass

    # Final fallback: print to stderr for debugging
    print(f"[cc-pomodoro] {title}: {message}", file=sys.stderr)


def _notify_linux(title: str, message: str) -> None:
    if shutil.which("notify-send"):
        subprocess.Popen(
            ["notify-send", title, message],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return
    raise FileNotFoundError("notify-send not found")


def _notify_macos(title: str, message: str) -> None:
    # terminal-notifier provides richer notifications
    if shutil.which("terminal-notifier"):
        subprocess.Popen(
            ["terminal-notifier", "-title", title, "-message", message],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return

    # Fallback to osascript native notification
    if shutil.which("osascript"):
        safe_title = title.replace('"', '\\"')
        safe_message = message.replace('"', '\\"')
        script = f'display notification "{safe_message}" with title "{safe_title}"'
        subprocess.Popen(
            ["osascript", "-e", script],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return
    raise FileNotFoundError("No notification command found on macOS")


def _notify_windows(title: str, message: str) -> None:
    # Escape special PowerShell characters in double-quoted strings
    safe_title = title.replace('"', '`"').replace('$', '`$')
    safe_message = message.replace('"', '`"').replace('$', '`$')

    # Method 1: PowerShell WinRT toast (requires AppUserModelID; may silently fail)
    winrt_script = (
        f'[Windows.UI.Notifications.ToastNotificationManager,'
        f' Windows.UI.Notifications, ContentType = WindowsRuntime] > $null; '
        f'$template = [Windows.UI.Notifications.ToastNotificationManager]::'
        f'GetTemplateContent([Windows.UI.Notifications.ToastTemplateType]::'
        f'ToastText02); '
        f'$textNodes = $template.GetElementsByTagName("text"); '
        f'$textNodes.Item(0).AppendChild('
        f'$template.CreateTextNode("{safe_title}")) > $null; '
        f'$textNodes.Item(1).AppendChild('
        f'$template.CreateTextNode("{safe_message}")) > $null; '
        f'$toast = [Windows.UI.Notifications.ToastNotification]::new($template); '
        f'[Windows.UI.Notifications.ToastNotificationManager]::'
        f'CreateToastNotifier().Show($toast)'
    )

    startupinfo = subprocess.STARTUPINFO()
    startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
    startupinfo.wShowWindow = 0  # SW_HIDE

    try:
        subprocess.Popen(
            ["powershell", "-NoProfile", "-NonInteractive", "-Command", winrt_script],
            startupinfo=startupinfo,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return
    except Exception:
        pass

    # Method 2: Fallback to msg.exe (available on all Windows editions)
    try:
        subprocess.Popen(
            ["msg", "*", f"{title}: {message}"],
            startupinfo=startupinfo,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return
    except Exception:
        pass

    # Method 3: COM Popup via PowerShell (always works)
    try:
        popup_script = (
            f'(New-Object -ComObject WScript.Shell).Popup('
            f'"{safe_message}", 5, "{safe_title}", 0x40)'
        )
        subprocess.Popen(
            ["powershell", "-NoProfile", "-NonInteractive", "-Command", popup_script],
            startupinfo=startupinfo,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return
    except Exception:
        pass

    raise RuntimeError("All Windows notification methods failed")
